Mathlib/Util/CompileInductive.lean

```diff
@@ -0,0 +1,179 @@
+      let pf := .app pf <| ← mkLambdaFVars xs[rv.getFirstIndexIdx:] <| ← mkEq body old new
+      let minor := xs[rv.getFirstMinorIdx]!
+      let pf := .app pf <| ← forallTelescope (← inferType minor) λ ys body' => do
+        let pf' ← mkEqRefl body' <| mkAppN minor ys
+        mkLambdaFVars ys pf'
+      let pf := .app pf xs[rv.getMajorIdx]!
+      mkFunExts' xs pf (body, old, new)
+  }
+  Compiler.CSimp.add name .global
+  compileDefn <| ← getConstInfoDefn <| mkRecOnName iv.name
+
+/--
+Compile the recursor for `iv`.
+-/
+def compileInductive (iv : InductiveVal) : TermElabM Unit := do
+  let rv ← getConstInfoRec <| mkRecName iv.name
+  if ← isProp rv.type then
+    logWarning m!"not compiling {rv.name}"
+    return
+  unless rv.numMotives == 1 do
+    throwError "mutual/nested inductives unsupported"
+  if iv.type.getForallBody.isProp && !iv.isRec && iv.numCtors == 1 && iv.numIndices == 0 then
+    compilePropStruct iv rv
+    return
+  let levels := rv.levelParams.map .param
+  let name ← mkFreshUserName rv.name
+  addAndCompile <| .mutualDefnDecl [{ rv with
+    name
+    value := ← forallTelescope rv.type λ xs body => do
+      let val := .const (mkCasesOnName iv.name) levels
+      let val := mkAppN val xs[:rv.numParams]
+      let val := .app val <| ← mkLambdaFVars xs[rv.getFirstIndexIdx:] body
+      let val := mkAppN val xs[rv.getFirstIndexIdx:]
+      let val := mkAppN val <| rv.rules.toArray.map λ rule =>
+        .beta (replaceConst rv.name name rule.rhs) xs[:rv.getFirstIndexIdx]
+      mkLambdaFVars xs val
+    hints := .opaque
+    safety := .partial
+  }]
+  let old := .const rv.name levels
+  let new := .const name levels
+  let name ← mkFreshUserName <| rv.name.str "eq"
+  addDecl <| .mutualDefnDecl [{
+    name
+    levelParams := rv.levelParams
+    type := ← mkEq rv.type old new
+    value := ← forallTelescope rv.type λ xs body => do
+      let pf := .const rv.name (.zero :: levels.tail!)
+      let pf := mkAppN pf xs[:rv.numParams]
+      let old := mkAppN old xs
+      let new := mkAppN new xs
+      let motive ← mkLambdaFVars xs[rv.getFirstIndexIdx:] <| ← mkEq body old new
+      let pf := .app pf motive
+      let pf := mkAppN pf <| ← rv.rules.toArray.zip xs[rv.getFirstMinorIdx:] |>.mapM λ (rule, minor) => do
+        forallTelescope ((← inferType minor).replaceFVar xs[rv.numParams]! motive) λ ys _ => do
+          let minor := mkAppN minor ys[:rule.nfields]
+          let pf' ← mkEqRefl (← inferType minor) minor
+          let pf' ← ys[rule.nfields:].foldlM (λ pf' y => do mkCongr pf' (← mkFunExts y)) pf'
+          mkLambdaFVars ys pf'
+      let pf := mkAppN pf xs[rv.getFirstIndexIdx:]
+      mkFunExts' xs pf (body, old, new)
+    hints := .opaque
+    safety := .partial
+  }]
+  Compiler.CSimp.add name .global
+  for aux in [mkRecOnName iv.name, mkBRecOnName iv.name] do
+    if let some (.defnInfo dv) := (← getEnv).find? aux then
+      compileDefn dv
+
+/--
+`#compile inductive i` compiles the recursor for `i`.
```

Comment:
This one, I do know what it's for. Perhaps:
```
`#compile inductive i` creates compiled code for the recursor for `i`,
so that it can be used in a definition without having to mark the definition as `noncomputable`.
```